src/v/storage/tiered_storage.cc

```diff
@@ -0,0 +1,516 @@
+partition_downloader::find_matching_partition_manifests(
+  cloud_storage::topic_manifest& manifest) {
+    // TODO: use only selected prefixes
+    auto topic_rev = manifest.get_revision();
+    std::vector<cloud_storage::remote_manifest_path> all_manifests;
+    auto obj_iter = [this, &all_manifests, topic_rev](
+                      const ss::sstring& key,
+                      std::chrono::system_clock::time_point,
+                      size_t,
+                      const ss::sstring&) {
+        std::filesystem::path path(key);
+        auto res = cloud_storage::get_manifest_path_components(path);
+        if (
+          res.has_value() && same_ntp(*res, _ntpc.ntp())
+          && res->_rev >= topic_rev) {
+            vlog(_ctxlog.debug, "Found matching manifest path: {}", *res);
+            all_manifests.emplace_back(
+              cloud_storage::remote_manifest_path(std::move(path)));
+        }
+        return ss::stop_iteration::no;
+    };
+    auto res = co_await _remote->list_objects(
+      obj_iter, _bucket, std::nullopt, /*TODO: use std::nullopt*/ 2, _rtcnode);
+    if (res == cloud_storage::download_result::success) {
+        co_return all_manifests;
+    }
+    auto key = manifest.get_manifest_path();
+    throw missing_partition_exception(key);
+}
+
+static ss::future<ss::output_stream<char>>
+open_output_file_stream(const std::filesystem::path& path) {
+    auto file = co_await ss::open_file_dma(
+      path.native(), ss::open_flags::rw | ss::open_flags::create);
+    auto stream = co_await ss::make_file_output_stream(std::move(file));
+    co_return std::move(stream);
+}
+
+ss::future<std::filesystem::path> partition_downloader::download_file(
+  const cloud_storage::segment_name& target,
+  const cloud_storage::manifest& manifest,
+  const std::filesystem::path& prefix) {
+    auto remote_location = manifest.get_remote_segment_path(target);
+    vlog(
+      _ctxlog.info,
+      "Downloading segment {} -> {} -> {}",
+      target,
+      remote_location,
+      prefix.string());
+    auto stream = [this, prefix, target, remote_location](
+                    uint64_t len,
+                    ss::input_stream<char> in) -> ss::future<uint64_t> {
+        auto path_or_fname = std::filesystem::path(target());
+        auto localpath = prefix / path_or_fname.filename();
+        vlog(
+          _ctxlog.info,
+          "Copying s3 path {} to local location {}",
+          remote_location,
+          localpath.string());
+        co_await ss::recursive_touch_directory(prefix.string());
+        auto fs = co_await open_output_file_stream(localpath);
+        co_await ss::copy(in, fs);
+        co_await fs.flush();
+        co_await fs.close();
+        co_return len;
+    };
+
+    auto result = co_await _remote->download_segment(
+      _bucket, target, manifest, stream, _rtcnode);
+
+    if (result != cloud_storage::download_result::success) {
+        // The individual segment might be missing for varios reasons but
```

Comment:
For a missing segment (a true 404) we should proceed, but this looks like it's ignoring *all* errors on download.

If someone has e.g. S3 connectivity issues, or a bad auth key, we should probably fail rather than giving them an empty topic, or a topic with gaps in?

Reply:
or is being rate limited by AWS